tests/integration_tests/test_hyperopt_ray_horovod.py

```diff
@@ -106,7 +106,7 @@ def mock_storage_client(path):
 
 
 # TODO ray: replace legacy mode when Ray Train supports placement groups
-RAY_BACKEND_KWARGS = {"processor": {"parallelism": 4}}
+RAY_BACKEND_KWARGS = {"processor": {"parallelism": 1}}
```

Comment:
We reduce `parallelism` in this workaround to ensure that Ray Datasets doesn't reserve all available CPUs ahead of the other trials being scheduled.

Before this change, all CPUs for the `train_fn` of each trial were scheduled up front by the Tuner. In this change, only the dummy `hyperopt_resources` are scheduled by the Tuner. The inner Tuners then schedule CPUs ad-hoc as they are called and executed by each trial.

The danger with this is in its interaction with Ray Datasets, which grabs resources opportunistically. If an inner Tuner is scheduled and its Ray Datasets tasks grab the remaining CPUs, other trials may be prevented from starting, causing the test to double in duration (since some trials are executed in sequence instead of all at once).

Setting `parallelism` to 1 here ensures that the number of CPUs requested by Ray Datasets is limited to 1 per trial.

Reply:
Right– so what we're seeing is not hanging behavior– instead what we are seeing is that the first `n` trials scheduled have Ray Datasets tasks that then take up the resources in the cluster. Eventually they let the resources go, at which point the `n+1`th trial begins executing. This is typically fine; however, for the sake of unit tests, this causes the test to take much longer, due to the fact that all `n+1` trials cannot execute exactly in parallel.

Reply:
Prior to this change, the trial driver would reserve the resources for all `n+1` trials at the onset. However, since there is this trial-within-trial scenario, only the top-level resources for the trial function (i.e. `hyperopt_resources`) are reserved at the onset. The `HorovodTrainer` resources (i.e. CPU/GPU) are reserved as the trial functions each hit that point in the code path.

Reply:
This also seems like a valuable comment to add to the code!